clientconn.go

```diff
@@ -133,12 +140,35 @@ func (cc *ClientConn) resetTransport(closeTransport bool) error {
 		if closeTransport {
 			t.Close()
 		}
-		newTransport, err := transport.NewClientTransport(cc.dopts.protocol, cc.target, cc.dopts.authOptions)
+		// Adjust timeout for the current try.
+		dopts := cc.dopts
```

Comment:
no need to copy the entire set of options. Only the timeout is relevant here. You can also collapse these two conditionals a bit.

```go
timeout := cc.dopts.Timeout
if timeout > 0 {
  timeout -= time.Since(start)
}
if timeout < 0 {
  ...
}
```